java/src/main/java/com/adhoc/flight/client/AdhocFlightClient.java

```diff
@@ -207,47 +223,71 @@ public FlightStream getStream(FlightInfo flightInfo, CallOption... options) {
    * @param headerCallOption client properties to execute provided SQL query with.
-   * @param fileToSaveTo the file to which the binary data of the resulting {@link VectorSchemaRoot}
-   *        should be saved.
-   * @param printToConsole whether or not the query results should be printed to the console.
+   * @param fileToSaveTo     the file to which the binary data of the resulting {@link VectorSchemaRoot}
+   *                         should be saved.
+   * @param printToConsole   whether the query results should be printed to the console.
    * @throws Exception if an error occurs during query execution.
    */
-  public void runQuery(String query,
-          HeaderCallOption headerCallOption,
-          File fileToSaveTo, boolean printToConsole) throws Exception {
+  public void runQuery(final String query,
+                       final HeaderCallOption headerCallOption,
+                       final @Nullable File fileToSaveTo,
+                       final boolean printToConsole) throws Exception {
 
     final FlightInfo flightInfo = getInfo(query, bearerToken, headerCallOption);
 
-    try (FlightStream stream = getStream(flightInfo, bearerToken, headerCallOption)) {
-      if (!stream.next()) {
-        return;
-      }
-
-      try (VectorSchemaRoot root = stream.getRoot()) {
+    try (final FlightStream flightStream = getStream(flightInfo, bearerToken, headerCallOption)) {
+      try (final VectorSchemaRoot allBatchesInRoot = unifyBatchesIntoSingleRoot(flightStream)) {
         if (printToConsole) {
-          QueryUtils.printResults(root);
+          QueryUtils.printResults(allBatchesInRoot);
         }
         if (fileToSaveTo != null) {
-          QueryUtils.writeToBinaryFile(root, fileToSaveTo);
+          try (final OutputStream outputStream = new FileOutputStream(fileToSaveTo)) {
+            outputRootBinaryDataToStream(allBatchesInRoot, outputStream);
+          }
         }
       }
     }
   }
 
   /**
-   * Make FlightRPC requests to the Dremio Flight Server Endpoint to retrieve results of the
-   * provided SQL query.
+   * Unifies all batches from the provided {@code flightStream} onto a new {@link VectorSchemaRoot}.
    *
-   * @param query            the SQL query to execute.
-   * @param headerCallOption client properties to execute provided SQL query with.
-   * @param fileToSaveTo     the file to which the binary data of the resulting {@link VectorSchemaRoot}
-   *                         should be saved.
-   * @throws Exception if an error occurs during query execution.
+   * @param flightStream the {@link FlightStream} instance from which to fetch the batches to unify.
+   * @return a new root.
    */
-  public void runQuery(String query, HeaderCallOption headerCallOption, File fileToSaveTo) throws Exception {
-    runQuery(query, headerCallOption, fileToSaveTo, false);
+  static VectorSchemaRoot unifyBatchesIntoSingleRoot(final FlightStream flightStream) {
```

Comment:
This seems like this won't scale well with large result sets. Won't this wind up storing everything in memory at once?

Reply:
Fixed.